vike/shared/page-configs/Config.ts

```diff
@@ -78,7 +81,11 @@ type ConfigBuiltIn = {
    *
    *  https://vike.dev/onBeforeRender
    */
-  onBeforeRender?: Function | ImportString | null
+  onBeforeRender?:
+    | ((pageContext: PageContextServer) => OptionalPromise<{ pageContext: Record<string, any> }>)
+    | ((pageContext: PageContextClient) => OptionalPromise<{ pageContext: Record<string, any> }>)
+    | ImportString
+    | null
```

Comment:
@brillout better?

Reply:
That's what I had in mind indeed. Interesting idea to have a union: I wonder whether this addresses the issue when a user changes the `env` of `onBeforeRender()` 🤔

Reply:
Thinking, maybe we could even do this.

```diff
- OptionalPromise<{ pageContext: Record<string, any> }>)
+ OptionalPromise<{ pageContext: Vike.PageContext }>)
```

Reply:
> I wonder whether this addresses the issue when a user changes the `env` of `onBeforeRender()`

oh am I missing something more from your previous points? I re-read them but I'm not sure which other issue there could be. The user can now freely do either

```
const onBeforeRender: Config['onBeforeRender'] = async (pageContext: PageContextServer) => {
```

or

```
const onBeforeRender: Config['onBeforeRender'] = async (pageContext: PageContextClient) => {
```

So any `env` is fine. Am I missing something?

However they can't do

```
const onBeforeRender: Config['onBeforeRender'] = async (pageContext: PageContextServer | PageContextClient) => {
```

so I need to add that to the union.



> Thinking, maybe we could even do this.
> 
> ```diff
> - OptionalPromise<{ pageContext: Record<string, any> }>)
> + OptionalPromise<{ pageContext: Vike.PageContext }>)
> ```

Let me give it a try

Reply:
> However they can't do
> 
> ```
> const onBeforeRender: Config['onBeforeRender'] = async (pageContext: PageContextServer | PageContextClient) => {
> ```
> 
> so I need to add that to the union.

Actually this works already, because `PageContextClient | PageContextServer` and `PageContextClient` happen to be the same thing already, as `PageContextClient` is already `<something> | PageContextServer`. So that's covered.

> Thinking, maybe we could even do this.
> 
> ```diff
> - OptionalPromise<{ pageContext: Record<string, any> }>)
> + OptionalPromise<{ pageContext: Vike.PageContext }>)
> ```

I would have to make `Page` and `config` optional at https://github.com/vikejs/vike/blob/main/examples/react-full-v1/renderer/PageContext.ts , so `onBeforeRender()` isn't forced to put them in its return value. It doesn't feel quite right though, for example in `vike-react` we don't make `Page` optional in `Vike.PageContext`, which we use as parameter of `onRenderHtml()`. I see two options:

1. we're happy with just `Record` as it is for now
2. we go for something like

```
  namespace Vike {
    interface PageContextAdditionalData {
      title?: string
      abortReason?: string
    }
  }
```

and `Vike.PageContext` is defined based on `Vike.PageContextAdditionalData`. And `onBeforeRender()` is expected to return `Vike.PageContextAdditionalData`. What's your feeling on this?

Reply:
Or how about [`Partial<Vike.PageContext>`](https://www.typescriptlang.org/docs/handbook/utility-types.html#partialtype)?

> `Vike.PageContext` is defined based on `Vike.PageContextAdditionalData`

AFAICT it would work indeed. It's an intersting idea to differentiate between refining existing `PageContext` and extending it.

I've a slight preference for `Partial<Vike.PageContext>` because it's more minimal (feel free to disagree). I'm curious to see whether they'll be more use cases for `Vike.PageContextAdditionalData` 👀.

> `PageContextClient` is already `<something> | PageContextServer`

It's actually slightly different: conceptually `PageContextClient` needs to be `Partial<PageContextServer>` because of `passToClient`. I've also made changes recently that increases the difference. So I think we do need the union `PageContextClient | PageContextServer`. (We can also use `PageContext` as `PageContext = PageContextClient | PageContextServer`.)

AFAICT everything will work then. Very neat.

One little drawback is that if the user doesn't type his hook:

```ts
const onBeforeRender: Config['onBeforeRender'] = async (pageContext) => {
```

then, I guess, TypeScript will set `pageContext` to `PageContextClient | PageContextServer`? Which doesn't align with the default `env` that is `server-only`. I guess we can tell users to set `pageContext` to `PageContextServer`. I can't think of a solution that sets `pageContext` to `PageContextServer` by default while allowing users to set it to `PageContextServer | PageContextClient` 🤔.

Let me know if something's not clear in my response!

Reply:
I didn't know `Partial<>`, all clear, thanks a lot :pray: I'll update the PR.